[ValueTracking] Let ComputeKnownSignBits handle (shl (zext X), C)

bjope opened the pull request:
Add simple support for looking through a zext when doing ComputeKnownSignBits for shl. This is valid for the case when all extended bits are shifted out, because then the number of sign bits can be found by analysing the zext operand.

The solution here is simple as it only handle a single zext (not passing remaining left shift amount during recursion). It could be possible to generalize this in the future by for example passing an 'OffsetFromMSB' parameter to ComputeNumSignBitsImpl, telling it to calculate number of sign bits starting at some offset from the most significant bit.

bjope commented:
SelectionDAG variant is here: https://github.com/llvm/llvm-project/pull/97695

nikic reviewed (commented):
This looks reasonable to me. Can you replace the unit tests with lit tests please?

bjope commented:
> This looks reasonable to me. Can you replace the unit tests with lit tests please?

I don't really know how to make good lit tests for ComputeNumSignBits.
All I've got is something like this
```
; RUN: opt -passes=instcombine -S < %s | FileCheck %s

define i16 @numsignbits_shl_zext(i8 %a) {
  %nsb6 = ashr i8 %a, 6
  %zext = zext i8 %nsb6 to i16
  %nsb4 = shl i16 %zext, 10
  %or = or i16 %nsb4, 1
  ; With four sign bits this should be nsw.
  %shl = add i16 %nsb4, 4096
  ret i16 %shl
}
```
to show that the `add` will get a `nsw`. But it far from verifies that ComputeNumSignBits actually is returning the correct value. Isn't it much easier to verify the correct behavior via the unit test?
Maybe I should add a lit test like the one above as a complement, to show that it is possible to trigger this via a lit test as well (even though it doesn't validate that the calculation in ComputeNumSignBits is correct).

goldsteinn commented on llvm/unittests/Analysis/ValueTrackingTest.cpp:
```diff
@@ -697,6 +697,36 @@ TEST_F(ValueTrackingTest, ComputeNumSignBits_PR32045) {
+                "  %A = shl i32 %c, 30\n"
+                "  ret i32 %A\n"
+                "}\n");
+  EXPECT_EQ(ComputeNumSignBits(A, M->getDataLayout()), 1u);
```
Probably should have a test (here or in lit) where `ShAmt < 24`

bjope commented:
Replaced unit test case with a lit based test (thanks to @goldsteinn for great tip about the and+add fold!).

dtcxzyw commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -3757,11 +3757,21 @@ static unsigned ComputeNumSignBitsImpl(const Value *V,
+        // all extended bits are shifted out.
+        if (match(U->getOperand(0), m_ZExt(m_Value(X))) &&
+            ShAmt->uge(TyBits - X->getType()->getScalarSizeInBits())) {
+          Tmp = ComputeNumSignBits(X, Depth + 1, Q);
```
Regression: https://alive2.llvm.org/ce/z/tcsmgq
Now `src` cannot be folded into `tgt`.
```
define i32 @src(i32 %add.i.i370) {
   %shr.i.i = ashr i32 %add.i.i370, 5
   %conv.i.i371 = zext i32 %shr.i.i to i64
   %shl.i.i = shl i64 %conv.i.i371, 34
    %conv6.i = ashr exact i64 %shl.i.i, 32
   %conv.i41.i = trunc nsw i64 %conv6.i to i32
   ret i32 %conv.i41.i
}

define i32 @tgt(i32 %add.i.i370) {
  %121 = ashr i32 %add.i.i370, 3
   %122 = and i32 %121, -4
   ret i32 %122
}
```

bjope replied:
I think it is related to this piece of code in InstCombinerImpl::SimplifyDemandedUseBits:

```
  case Instruction::AShr: {
    unsigned SignBits = ComputeNumSignBits(I->getOperand(0), Depth + 1, Q.CxtI);
    ...
      // If the input sign bit is known to be zero, or if none of the top bits
      // are demanded, turn this into an unsigned shift right.
      assert(BitWidth > ShiftAmt && "Shift amount not saturated?");
      APInt HighBits(APInt::getHighBitsSet(
                         BitWidth, std::min(SignBits + ShiftAmt - 1, BitWidth)));
      if (Known.Zero[BitWidth-ShiftAmt-1] ||
          !DemandedMask.intersects(HighBits)) {
        BinaryOperator *LShr = BinaryOperator::CreateLShr(I->getOperand(0),
                                                          I->getOperand(1));
        LShr->setIsExact(cast<BinaryOperator>(I)->isExact());
        LShr->takeName(I);
        return InsertNewInstWith(LShr, I->getIterator());
      }
```

And I do not fully undertstand the logic for the HigBits computation.
Given` %conv6.i = ashr exact i64 %shl.i.i, 32` , I think that BitWidth is 64. And ShifAmt is 32.
So the code is doing
`   APInt::getHighBitsSet(64, std::min(SignBits + 32 - 1, 64)`
And if we know nothing about sign bits we will set the 32 high bits. But the more we know about sign bits, the more bits will be set. With this patch we SignBits==4 so we get more bits set in HighBits.

But then the check `!DemandedMask.intersects(HighBits)` is less likely to be fulfilled. Because the intersects would be more likely to happen if we are able to compute more sign bits. So the better ComputeNumSignBits analysis we have, the less likely that we apply the optimization converting the AShr to LShr. That doesn't make sense to me!

bjope replied:
Seems to originate from this patch back in 2017: https://github.com/llvm/llvm-project/commit/22178dd33b3460207b8b1dd109339524e50feab5
So the idea actually was to avoid ASHR->LSHR rewrites in certain situations (based on number of known sign bits).

Unfortunately there are no good motivating test cases. Even the test case added in that commit would still pass if reverting the change in SimplifyDemandedUseBits.

bjope replied:
Created https://github.com/llvm/llvm-project/pull/99155 to avoid such regressions.

bjope replied:
Rebased branch on top of https://github.com/llvm/llvm-project/commit/b8c4c58ecf186dd91f40bdff4d1bdad403435789 to show that the regression no longer can be seen.

bjope committed:
[ValueTracking] Let ComputeKnownSignBits handle (shl (zext X), C) (#97693)

Add simple support for looking through a zext when doing
ComputeKnownSignBits for shl. This is valid for the case when
all extended bits are shifted out, because then the number of sign
bits can be found by analysing the zext operand.

The solution here is simple as it only handle a single zext (not
passing remaining left shift amount during recursion). It could be
possible to generalize this in the future by for example passing an
'OffsetFromMSB' parameter to ComputeNumSignBitsImpl, telling it to
calculate number of sign bits starting at some offset from the most
significant bit.

nikic reviewed (approved):
LGTM

dtcxzyw reviewed (approved):
LGTM. Thank you!
